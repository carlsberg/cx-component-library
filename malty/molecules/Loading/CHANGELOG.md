# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2024-05-31)

### Bug Fixes

- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-254:** loading molecule ([#211](https://github.com/CarlsbergGBS/cx-component-library/issues/211)) ([52b9bd4](https://github.com/CarlsbergGBS/cx-component-library/commit/52b9bd43d3dc59a6d65afdcf2c236d1d669dee98))
- **DSM-337:** button color ([#295](https://github.com/CarlsbergGBS/cx-component-library/issues/295)) ([31349ca](https://github.com/CarlsbergGBS/cx-component-library/commit/31349cadbbef7d0d5915b150349a3f7c90fc01ff))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- **loading-overlay:** create component ([#856](https://github.com/CarlsbergGBS/cx-component-library/issues/856)) ([32e0597](https://github.com/CarlsbergGBS/cx-component-library/commit/32e05974e0240b4d317183900469b2abf0a957e3))
- **loading:** [DSM-930] add tests ([#919](https://github.com/CarlsbergGBS/cx-component-library/issues/919)) ([668aa2a](https://github.com/CarlsbergGBS/cx-component-library/commit/668aa2ab847506d6fad730301edec1423fa24a5f))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- **table:** [DSM-956] add loading overlay ([#867](https://github.com/CarlsbergGBS/cx-component-library/issues/867)) ([bbcf090](https://github.com/CarlsbergGBS/cx-component-library/commit/bbcf09087014c96f56cfe6b97fddf15999584247))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
