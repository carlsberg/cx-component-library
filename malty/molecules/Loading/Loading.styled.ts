import styled, { keyframes } from 'styled-components';

const rotate = keyframes`
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
`;

const fadeIn = keyframes`
    from {
      opacity: 0;
      transform: rotate(270deg);
    }
    to {
      opacity: 1;
      transform: rotate(360deg);
    }
`;

export const StyledLoadingContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${({ theme }) => theme.sizes['3xs'].value};
  align-items: center;
  justify-content: center;
  text-align: center;
  padding: ${({ theme }) => theme.sizes['3xs'].value};
  letter-spacing: 0;
  color: ${({ theme }) => theme.colors.colours.support[60].value};
  font-size: ${({ theme }) => theme.typography.desktop.text.small_bold['font-size'].value};
  font-weight: ${({ theme }) => theme.typography.desktop.text.small_bold['font-weight'].value};
  line-height: ${({ theme }) => theme.typography.desktop.text.small_bold['line-height'].value};
  font-family: ${({ theme }) => theme.typography.desktop.text.small_bold['font-family'].value};
`;

export const StyledLoading = styled.div<{
  size: string;
}>`
  display: flex;
  svg {
    &.fade-in {
      animation: ${fadeIn} 1s;
    }
    &.spinning {
      animation: ${rotate} 2s linear infinite;
    }
    height: ${({ size }) => size};
    width: ${({ size }) => size};
  }
`;
