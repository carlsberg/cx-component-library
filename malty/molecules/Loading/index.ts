export { Loading } from './Loading';
export { LoadingSize, LoadingStatus } from './Loading.types';
export type { LoadingProps } from './Loading.types';
