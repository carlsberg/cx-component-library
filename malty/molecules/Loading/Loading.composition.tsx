import { MaltyThemeProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React from 'react';
import { Loading } from './Loading';
import { LoadingStatus, SizeTypes } from './Loading.types';

export const BasicCodeInput = () => (
  <MaltyThemeProvider theme="global">
    <Loading text="Loading..." size={SizeTypes.Medium} status={LoadingStatus.Pending} />
  </MaltyThemeProvider>
);
