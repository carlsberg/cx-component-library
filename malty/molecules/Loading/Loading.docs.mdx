---
labels: ['react', 'typescript', 'icon']
description: 'A Icon component.'
---

import { MaltyThemeProvider } from '@carlsberg/malty.theme.malty-theme-provider';
import { Loading } from './Loading';
import { SizeTypes, LoadingStatus } from './Loading.types';

Icon example:

An simple Icon component with some text

Live Playground:

```js live
<MaltyThemeProvider theme="global">
  <Loading text={'Loading...'} size={SizeTypes.Large} status={LoadingStatus.Default} />
</MaltyThemeProvider>
```
