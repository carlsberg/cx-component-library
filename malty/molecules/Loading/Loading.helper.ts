import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { LoadingSize, LoadingStatus, UseLoadingStatusProps, UseLoadingStylesProps } from './Loading.types';

export const useLoadingStyles = ({ size }: UseLoadingStylesProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  const loadingStyles = {
    [LoadingSize.Small]: {
      iconSize: theme.sizes.m.value
    },
    [LoadingSize.Medium]: {
      iconSize: theme.sizes['2xl'].value
    }
  };

  return loadingStyles[size];
};

export const useLoadingStatus = ({ status }: UseLoadingStatusProps) => {
  if (!status) return LoadingStatus.Pending;
  return status;
};
