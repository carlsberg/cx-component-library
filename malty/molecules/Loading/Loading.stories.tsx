import { Meta, Story } from '@storybook/react';
import React from 'react';
import { Loading } from './Loading';
import { LoadingProps, LoadingStatus, SizeTypes } from './Loading.types';

export default {
  title: 'Molecules/Loading',
  component: Loading,
  parameters: {
    importObject: 'Loading',
    importPath: '@carlsberggroup/malty.molecules.loading'
  },
  argTypes: {
    text: { control: 'text' },
    size: {
      options: Object.values(SizeTypes),
      control: {
        type: 'select'
      }
    },
    status: {
      options: Object.values(LoadingStatus),
      control: {
        type: 'select'
      }
    }
  }
} as Meta;

const Template: Story<LoadingProps> = ({ text, size, status }: LoadingProps) => (
  <Loading text={text} size={size} status={status} />
);

export const Main = Template.bind({});
Main.args = {
  size: SizeTypes.Medium,
  text: 'Loading...',
  status: LoadingStatus.Pending
};
