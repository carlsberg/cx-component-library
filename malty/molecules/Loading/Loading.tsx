import {
  ProgressSpinner,
  ProgressSpinnerColor,
  ProgressSpinnerStatus
} from '@carlsberggroup/malty.atoms.progress-spinner';
import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useEffect, useState } from 'react';
import { StyledLoading, StyledLoadingContainer } from './Loading.styled';
import { LoadingProps, LoadingSize, LoadingStatus } from './Loading.types';

export function Loading({
  text,
  size = LoadingSize.Small,
  status = LoadingStatus.Pending,
  dataQaId,
  negative = false,
  color = ProgressSpinnerColor.DigitalBlack
}: LoadingProps) {
  const theme = defaultTheme;

  const [progressStatus, setProgressStatus] = useState<ProgressSpinnerStatus>(ProgressSpinnerStatus.Pending);
  const [iconSize, setIconSize] = useState(theme.sizes.m.value);

  useEffect(() => {
    switch (size) {
      case LoadingSize.Medium: {
        setIconSize(theme.sizes['2xl'].value);

        break;
      }
      default: {
        setIconSize(theme.sizes.m.value);

        break;
      }
    }
  }, [size, theme]);

  useEffect(() => {
    switch (status) {
      case LoadingStatus.Success: {
        setProgressStatus(ProgressSpinnerStatus.Success);
        break;
      }
      case LoadingStatus.Failure: {
        setProgressStatus(ProgressSpinnerStatus.Failure);
        break;
      }
      default: {
        setProgressStatus(ProgressSpinnerStatus.Pending);
        break;
      }
    }
  }, [status]);

  return status ? (
    <StyledLoadingContainer data-testid={`${dataQaId}`} size={size} theme={theme}>
      <StyledLoading
        size={iconSize}
        className={`${status === LoadingStatus.Pending ? 'spinning' : 'fade-in'} ${status}`}
      >
        <ProgressSpinner color={color} negative={negative} dataQaId={`${dataQaId}`} status={progressStatus} />
      </StyledLoading>

      {text && (
        <Text dataQaId={`${dataQaId}-label`} textStyle={TextStyle.SmallBold} color={TextColor.Support60}>
          {text}
        </Text>
      )}
    </StyledLoadingContainer>
  ) : null;
}
