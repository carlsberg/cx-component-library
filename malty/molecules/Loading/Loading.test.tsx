import { jsonRenderer, render, screen } from '@carlsberggroup/malty.utils.test';
import React from 'react';
import { Loading } from './Loading';
import { LoadingStatus, SizeTypes } from './Loading.types';

describe('molecule loading', () => {
  it('matches snapshot', () => {
    const view = jsonRenderer(<Loading text="Loading..." size={SizeTypes.Medium} status={LoadingStatus.Pending} />);
    expect(view).toMatchSnapshot();
  });

  it('renders element with text', () => {
    render(<Loading text="Loading..." size={SizeTypes.Medium} status={LoadingStatus.Pending} />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders success icon', () => {
    render(<Loading size={SizeTypes.Medium} status={LoadingStatus.Success} />);
    expect(screen.getByTestId('svg-component')).toHaveClass(LoadingStatus.Success);
  });

  it('renders failure icon', () => {
    render(<Loading size={SizeTypes.Medium} status={LoadingStatus.Failure} />);
    expect(screen.getByTestId('svg-component')).toHaveClass(LoadingStatus.Failure);
  });
});
