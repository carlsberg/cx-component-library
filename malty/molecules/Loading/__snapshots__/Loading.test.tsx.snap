// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecule loading matches snapshot 1`] = `
<div
  className="sc-dkPtRN dAZxbL"
>
  <div
    className="sc-hKwDye ejyGmF"
    size="24"
  >
    <svg
      className="sc-gsDKAQ kyUyaQ spinning Pending"
      color="digital-black"
      data-testid="svg-component"
      size="24px"
      version="1.1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <g
        fillRule="evenodd"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2zm0 2a8 8 0 110 16 8 8 0 010-16z"
          fill="#ffffff"
        />
        <path
          d="M12 2c5.43 0 9.848 4.327 9.996 9.72L22 12h-2a8 8 0 00-7.75-7.996L12 4z"
        />
      </g>
    </svg>
  </div>
  Loading...
</div>
`;
